Create scraper for CourseBook data

CharlieMahana opened the pull request:
Closes #22.

WillieCubed commented on scraper/scraper.py:
```diff
@@ -0,0 +1,212 @@
+            'Accept-Encoding': 'gzip, deflate, br',
+            'Accept-Language': 'en-US,en;q=0.9',
+            'Connection': 'keep-alive',
+            'Cookie': 'PTGSESSID={}'.format(self.ptgsessid),
```
f-strings are cooler.
```suggestion
            'Cookie': f'PTGSESSID={self.ptgsessid}',
```

WillieCubed commented on scraper/scraper.py:
```diff
@@ -0,0 +1,212 @@
+        }
+
+        # send coursemonkey download request for json
+        # self.logger.info('Sending ReportMonkey query: {0}'.format(query))
```
Remove commented out code when merging to `develop`. If it's important, add it to another branch.

WillieCubed commented on scraper/scraper.py:
```diff
@@ -0,0 +1,212 @@
+        # obtain coursebook index page
+        self.logger.info('Getting CourseBook index page')
+        try:
+            r = requests.get('https://coursebook.utdallas.edu')
```
Could you extract the CourseBook URL to a top-level variable and replace it anywhere it is used?

WillieCubed commented on scraper/scraper.py:
```diff
@@ -0,0 +1,212 @@
+            return {}
+
+        # create reportmonkey downloadURL from downloadID
+        downloadURL = 'https://coursebook.utdallas.edu/reportmonkey/cb11-export/{0}/{0}/json'.format(downloadID)
```
```suggestion
        downloadURL = f'https://coursebook.utdallas.edu/reportmonkey/cb11-export/{downloadID}/{downloadID}/json'
```

WillieCubed commented on scraper/scraper.py:
```diff
@@ -0,0 +1,212 @@
+        self.logger.info('Started coursebook scraper')
+
+
+    def getSearchSpecifications(self) -> dict:
```
I'll note this now: stick to PEP 8 (the Python style guide) and use that convention for all of the variable/function/class names: https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles

WillieCubed commented on scraper/scraper.py:
```diff
@@ -0,0 +1,212 @@
+        }
+
+        # send CourseBook search query
+        # self.logger.info('Sending CourseBook query: {0}'.format(query))
```
Don't forget to remove this if it isn't necessary for functionality.

WillieCubed commented on scraper/scraper.py:
```diff
@@ -0,0 +1,212 @@
+    scraper = CoursebookScraper(os.environ['PTGSESSID'])
+
+    # load default config
+    with open('configs/{}'.format(os.environ['CONFIG_FILE']), 'r') as file:
```
```suggestion
    config_file_name = os.environ['CONFIG_FILE']
    with open(f'configs/{config_file_name}', 'r') as file:
```

WillieCubed commented on scraper/scraper.py:
```diff
@@ -0,0 +1,212 @@
+    result = scraper.downloadFromConfig()
+
+    # write results to data.json
+    with open('data/{}.json'.format(int(time.time())),'w') as file:
```
```suggestion
    with open(f'data/{int(time.time()}.json', 'w') as file:
```

WillieCubed commented on scraper/scraper.py:
```diff
@@ -0,0 +1,212 @@
+        try:
+            numItems = html.find_all('b')[0].__repr__().split('(')[1].split(')')[0].split(' ')[0] # this should probably be cleaned up
+            if numItems == 'no':
+                self.logger.info('0 results obtained from query: {0}'.format(query))
```
```suggestion
                self.logger.info(f'0 results obtained from query: {query}')
```

WillieCubed commented on scraper/scraper.py:
```diff
@@ -0,0 +1,212 @@
+        # format query list as a proper query string
+        query = 'action=search'
+        for specifier in queryList:
+            query += '&s[]={0}'.format(specifier)
```
```suggestion
            query += f'&s[]={specifier}'
```
Sidenote for future reference: I'm almost certain there's a library that does something like this, but this is fine.

WillieCubed commented on scraper/scraper.py:
```diff
@@ -0,0 +1,212 @@
+            )
+            html = BeautifulSoup(response.content, 'html.parser')
+        except:
+            self.logger.error('Failed to receive response from query: {0}'.format(query))
```
```suggestion
            self.logger.error(f'Failed to receive response from query: {query}')
```

WillieCubed commented on scraper/scraper.py:
```diff
@@ -0,0 +1,212 @@
+
+        # specify CourseBook request header
+        coursebookHeader = {
+            'Cookie': 'PTGSESSID={0}'.format(self.ptgsessid),
```
```suggestion
            'Cookie': f'PTGSESSID={self.ptgsessid}',
```

WillieCubed commented on scraper/scraper.py:
```diff
@@ -0,0 +1,212 @@
+                self.logger.info('0 results obtained from query: {0}'.format(query))
+                return {}
+        except:
+            self.logger.error('Failed to parse number of items returned from query: {0}'.format(query))
```
```suggestion
            self.logger.error(f'Failed to parse number of items returned from query: {query}')
```

WillieCubed commented on scraper/scraper.py:
```diff
@@ -0,0 +1,212 @@
+            else:
+                raise Exception()
+        except:
+            self.logger.error('Failed to parse download link to {0} item(s) from query: {1}'.format(numItems, query))
```
```suggestion
            self.logger.error(f'Failed to parse download link to {numItems} item(s) from query: {query}')
```

WillieCubed commented on scraper/scraper.py:
```diff
@@ -0,0 +1,212 @@
+            download = requests.get(downloadURL, headers=reportmonkeyHeader)
+            rawJSON = download.text
+        except:
+            self.logger.error('Failed to download JSON response for query: {0}'.format(query))
```
```suggestion
            self.logger.error(f'Failed to download JSON response for query: {query}')
```

WillieCubed commented on scraper/scraper.py:
```diff
@@ -0,0 +1,212 @@
+        try:
+            queryAsJSON = json.loads(rawJSON)
+        except:
+            self.logger.error('Failed to convert JSON to dict for query: {0}'.format(query))
```
```suggestion
            self.logger.error(f'Failed to convert JSON to dict for query: {query}')
```

WillieCubed reviewed (changes requested):
Some relatively minor changes to make, but nice job on the script.

CharlieMahana commented:
I fixed all the issues you mentioned and removed the hardcoded data file in dataToFirestore.py so that the file can be passed into the program through command line arguments

WillieCubed reviewed (changes requested):
Ah, there was a comment that was missed [here](https://github.com/UTDNebula/nebula-api/pull/22/commits/9526714b629d9ccbaa2a7a04f26c1dcae99ac451#r746878065). Additionally, add `.env` to the `.gitignore`.

CharlieMahana commented:
I removed the commented code block, finished f-string conversions, and added /scraper/.env to .gitignore. My only concern is that firebase credentials could accidentally be committed to the repository as the credentials file gets a new name every time we generate one. How do we want to solve this problem?

WillieCubed commented:
In the project set-up docs, just inform people to rename the file so it has the correct hardcoded name a that is how it is intended to work.

CharlieMahana commented:
@WillieCubed my most recent commit should mean everything has been taken care of and we're ready to merge. Let me know if there is anything that I missed.

WillieCubed reviewed (approved):
LGTM!
